executor/src/wasm_utils.rs

```diff
@@ -0,0 +1,221 @@
+			dispatch!($objectname, $( $name( $( $names : $params ),* ) $( -> $returns )* => $body ),*);
+		}
+		impl $( $pre ) + $structname {
+			signatures!($( $name( $( $params ),* ) $( -> $returns )* ),*);
+		}
+		impl $( $pre ) + $crate::wasm_utils::IntoUserDefinedElements for $structname {
+			fn into_user_defined_elements(&mut self) -> UserDefinedElements<$crate::wasm_utils::DummyUserError> {
+				$crate::wasm_utils::UserDefinedElements {
+					executor: Some(self),
+					globals: HashMap::new(),	// TODO: provide
+					functions: ::std::borrow::Cow::from(Self::SIGNATURES),
+				}
+			}
+		}
+	);
+}
+
+#[derive(Clone)]
+struct DummyUserFunctionExecutor;
+impl<E: UserError> interpreter::UserFunctionExecutor<E> for DummyUserFunctionExecutor {
+	fn execute(&mut self, _name: &str, _context: interpreter::CallerContext<E>) ->
+		result::Result<Option<interpreter::RuntimeValue>, interpreter::Error<E>>
+	{
+		unimplemented!()
+	}
+}
+
+pub trait AddModuleWithoutFullDependentInstance {
+	fn add_module_by_sigs(
+		&self,
+		name: &str,
+		module: Module,
+		functions: HashMap<&str, &'static [UserFunctionDescriptor]>,
+	) -> result::Result<Arc<interpreter::ModuleInstance<DummyUserError>>, interpreter::Error<DummyUserError>>;
+
+	fn params_with_external<'a, 'b: 'a>(&'b self, externals_name: &str, externals: &'a mut IntoUserDefinedElements) -> ExecutionParams<'a, DummyUserError>;
+}
+
+impl AddModuleWithoutFullDependentInstance for interpreter::ProgramInstance<DummyUserError> {
+	fn add_module_by_sigs(
+		&self,
+		name: &str,
+		module: Module,
+		functions: HashMap<&str, &'static [UserFunctionDescriptor]>
+	) -> result::Result<Arc<interpreter::ModuleInstance<DummyUserError>>, interpreter::Error<DummyUserError>> {
+		let mut dufe = vec![DummyUserFunctionExecutor; functions.len()];
+		let dufe_refs = dufe.iter_mut().collect::<Vec<_>>();
+		let fake_module_map = functions.into_iter()
+			.zip(dufe_refs.into_iter())
+			.map(|((dep_mod_name, functions), dufe)| {
+				let fake_module = Arc::new(
+					interpreter::env_native_module(
+						self.module(dep_mod_name).unwrap(), UserDefinedElements {
+							executor: Some(dufe),
+							globals: HashMap::new(),
+							functions: ::std::borrow::Cow::from(functions),
+						}
+					).unwrap()	// TODO: handle
+				);
+				let fake_module: Arc<interpreter::ModuleInstanceInterface<_>> = fake_module;
+				(dep_mod_name.into(), fake_module)
+			})
+			.collect::<HashMap<_, _>>();
+		self.add_module(name, module, Some(&fake_module_map))
+	}
+
+	fn params_with_external<'a, 'b: 'a>(&'b self, externals_name: &str, externals: &'a mut IntoUserDefinedElements) -> ExecutionParams<'a, DummyUserError> {
+		interpreter::ExecutionParams::with_external(
+			externals_name.into(),
+			Arc::new(
+				interpreter::env_native_module(
+					self.module(externals_name).unwrap(),	// TODO: handle
+					externals.into_user_defined_elements()
+				).unwrap()	// TODO: handle
+			)
+		)
+	}
+}
+
+#[macro_export]
+macro_rules! map {
```

Comment:
this macro name (among others in this module) is way too general to export. I recommend making these sub-invocations of the `impl_function_executor` macro.

Reply:
we have `vec!`. why not `map!`?